encoders/firebase-decoders-json/src/main/java/com/google/firebase/decoders/TypeToken.java

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders;
+
+import androidx.annotation.NonNull;
+
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.WildcardType;
+
+//TODO: implement hashCode(), equals(), and toString().
+public abstract class TypeToken<T> {
+
+  @NonNull
+  public static <T> TypeToken<T> of(@NonNull Safe<T> token) {
+    Type superclass = token.getClass().getGenericSuperclass();
+    if (superclass instanceof Class) {
+      throw new RuntimeException("Missing type parameters");
```

Comment:
pls throw a more specific exception, e.g. `IllegalArgumentException` is a good candidate in this case.